src/v/cluster/rm_stm.cc

```diff
@@ -416,20 +416,43 @@ ss::future<tx_errc> rm_stm::do_prepare_tx(
         co_return tx_errc::fenced;
     }
 
-    if (synced_term != etag) {
-        vlog(
-          clusterlog.warn,
-          "Can't prepare pid:{} - partition lost leadership current term: {} "
-          "expected term: {}",
-          pid,
-          synced_term,
-          etag);
-        // current partition changed leadership since a transaction started
-        // there is a chance that not all writes were replicated
-        // rejecting a tx to prevent data loss
-        co_return tx_errc::request_rejected;
-    }
+    /*
+        // -----
+        // This check _seems_ redundant along with the `expected` check below,
+       especially
+        // with these set of changes that enable a transaction to span multiple
+       terms.
+        //
+        // This `etag`/`expected_term` enforces a single term invariant for a
+       given
+        // transaction and data partition pair. Ideally this is also tracked in
+       the
+        // `expected` state which tracks all the inflight pids.
+
+        // During a crash `expected` state is cleared up, so we'd still
+       guarantee that we
+        // do not mark a lost write as `prepared`. However if checkpointing is
+       correctly
+        // done, a single transaction _can_ span multiple terms without any
+       issues.
 
+        // TODO: this needs more research and tighter validation with targeted
+        // chaos tests. Denis, FYI.
+
+        // -----
+
+
+        if (synced_term != etag) {
```

Comment:
I'm pretty sure you're right but it's dangerous to relax conditions before the release. We should  update mem_state.expected to store a term (actual at the begin_tx moment) along the tx_seq. Then when we transfer the state we'll be able to continue using the check for validating the prepare and then using the term of `receive_tx_leadership` for conditional replicate.

Reply:
Redid this by tracking the term in begin_tx. Although I'm not fully convinced that version is any better than commenting the check. Anyway check the code and we can discuss.

Reply:
I on the same page with you :)